File: VI-semester/JPP/L2/python/finite_field.py
class FiniteField:
    def __init__(self, prime_number, value):
        if self.is_prime(prime_number):
            self.characteristic = prime_number
        else:
            raise Exception("The received number in the constructor is NOT a prime number!")
        self.value = (value % self.characteristic + self.characteristic) % self.characteristic

    def is_prime(self, number):
        if number <= 1:
            return False
        if number == 2:
            return True
        if number % 2 == 0:
            return False
        for i in range(3, int(number ** 0.5) + 1, 2):
            if number % i == 0:
                return False
        return True

    def get_characteristic(self):
        return self.characteristic

    def get_value(self):
        return self.value

    # COMPARING OPERATORS
    def __eq__(self, other):
        return self.value == other.value

    def __ne__(self, other):
        return self.value != other.value

    def __le__(self, other):
        return self.value <= other.value

    def __ge__(self, other):
        return self.value >= other.value

    def __lt__(self, other):
        return self.value < other.value

    def __gt__(self, other):
        return self.value > other.value

    # ARITHMETIC OPERATORS
    def __add__(self, other):
        return FiniteField(self.characteristic, (self.value + other.value) % self.characteristic)

    def __sub__(self, other):
        result = (self.value - other.value) % self.characteristic
        if result < 0:
            result += self.characteristic
        return FiniteField(self.characteristic, result)

    def __mul__(self, other):
        first = self.value % self.characteristic
        second = other.value % self.characteristic
        result = first
        for i in range(1, second):
            result += first
            result = ((result % self.characteristic) + self.characteristic) % self.characteristic
        return FiniteField(self.characteristic, result)

    def extended_gcd(self, a, b):
        if a == 0:
            return [0, 1, b]

        gcd = self.extended_gcd(b % a, a)
        x, y = gcd[1] - (b // a) * gcd[0], gcd[0]

        return [x, y, gcd[2]]

    def mod_invert(self, a, m):
        gcd = self.extended_gcd(a, m)
        if gcd[2] != 1:
            # Modular inverse does not exist
            return -1
        else:
            return self.__mod__(gcd[0], m)

    def __truediv__(self, other):
        inv = self.mod_invert(other.value, other.characteristic)
        if inv == -1:
            raise ArithmeticError("Modular inverse does not exist!")
        inverse = FiniteField(self.characteristic, inv)
        return self.__mul__(inverse)

    def __mod__(self, left, right):
        return (left % right + self.characteristic) % self.characteristic

    # ASSIGNING OPERATORS
    def assign(self, other):
        return FiniteField(self.characteristic, other.value)

    def __iadd__(self, other):
        return FiniteField(self.characteristic, self.__add__(other).value)

    def __isub__(self, other):
        return FiniteField(self.characteristic, self.__sub__(other).value)

    def __imul__(self, other):
        return FiniteField(self.characteristic, self.__mul__(other).value)

    def __idiv__(self, other):
        return FiniteField(self.characteristic, self.__truediv__(other).value)

File: VI-semester/JPP/L2/java/FiniteField.java
public class FiniteField {
    private final int characteristic;
    private int value;
    public FiniteField(int primeNumber, int value) throws Exception {
        if (isPrime(primeNumber)) {
            characteristic = primeNumber;
        }
        else {
            throw new Exception("The received number in the constructor is NOT a prime number!");
        }
        this.value = (value % characteristic + characteristic) % characteristic;
    }

    private boolean isPrime(int number) {
        if (number <= 1) {
            return false;
        }
        if (number == 2) {
            return true;
        }
        if (number % 2 == 0) {
            return false;
        }
        for (int i = 3; i <= Math.sqrt(number); i += 2) {
            if (number % i == 0) {
                return false;
            }
        }
        return true;
    }

    public int getCharacteristic() {
        return characteristic;
    }
    public int getValue() { return value; }

    // ################### COMPARING OPERATORS ###################
    public boolean isEqual(FiniteField other) {
        return this.value == other.value;
    }

    public boolean isNotEqual(FiniteField other) {
        return this.value != other.value;
    }

    public boolean isLessOrEqual(FiniteField other) {
        return this.value <= other.value;
    }

    public boolean isGreaterOrEqual(FiniteField other) {
        return this.value >= other.value;
    }

    public boolean isLess(FiniteField other) {
        return this.value < other.value;
    }

    public boolean isGreater(FiniteField other) {
        return this.value > other.value;
    }

    // ################### ARITHMETIC OPERATORS ###################
    public FiniteField add(FiniteField other) throws Exception {
        return new FiniteField(characteristic, (this.value + other.value) % characteristic);
    }

    public FiniteField subtract(FiniteField other) throws Exception {
        int result = (this.value - other.value) % characteristic;
        if (result < 0) {
            result += characteristic;
        }
        return new FiniteField(characteristic, result % characteristic);
    }

    public FiniteField multiply(FiniteField other) throws Exception {
        int first = this.value % characteristic;
        int second = other.value % characteristic;
        int result = first;
        for(int i = 1; i < second; i++) {
            result += first;
            result = ((result % characteristic) + characteristic) % characteristic;
        }
        return new FiniteField(characteristic, result);
    }

    private int[] extended_gcd(int a, int b) {
        if (a == 0) {
            return new int[]{0, 1, b};
        }

        int[] gcd = extended_gcd(b % a, a);
        int x, y;
        x = gcd[1] - (b / a) * gcd[0];
        y = gcd[0];

        return new int[]{x, y, gcd[2]};
    }

    private int mod_invert(int a, int m) {
        int[] gcd = extended_gcd(a, m);

        if (gcd[2] != 1) {
            // Modular inverse does not exist
            return -1;
        } else {
            return modulo(gcd[0], m);
        }
    }

    public FiniteField divide(FiniteField other) throws Exception {
        int inv = mod_invert(other.value, other.characteristic);
        if (inv == -1) {
            throw new ArithmeticException("Modular inverse does not exist!");
        }
        FiniteField inverse = new FiniteField(characteristic, inv);
        return multiply(inverse);
        //inv = assign(inv);

        //return modulo(multiply(left, inv), characteristic);
    }

    public int modulo(int left, int right) {
        return (left % right + characteristic) % characteristic;
    }

    // ################### ASSIGNING OPERATORS ###################

    public FiniteField assign(FiniteField other) throws Exception {
        return new FiniteField(characteristic, other.value);
    }

    public FiniteField addAssign(FiniteField other) throws Exception {
        return new FiniteField(characteristic, add(other).value);
    }

    public FiniteField subtractAssign(FiniteField other) throws Exception {
        return new FiniteField(characteristic, subtract(other).value);
    }

    public FiniteField multiplyAssign(FiniteField other) throws Exception {
        return new FiniteField(characteristic, multiply(other).value);
    }

    public FiniteField divideAssign(FiniteField other) throws Exception {
        return new FiniteField(characteristic, divide(other).value);
    }
}
